Source/DafnyStandardLibraries/src/Std/Actions/Actions.dfy

```diff
@@ -0,0 +1,620 @@
+      && Produced() == Seq.MapPartialFunction(f, Consumed())
+    }
+
+    constructor(f: T -> R)
+      ensures Valid()
+      ensures this.f == f
+      ensures fresh(Repr)
+      ensures history == []
+    {
+      this.f := f;
+
+      history := [];
+      Repr := {this};
+    }
+
+    ghost predicate CanConsume(history: seq<(T, R)>, next: T)
+      requires CanProduce(history)
+      decreases height
+    {
+      f.requires(next)
+    }
+    ghost predicate CanProduce(history: seq<(T, R)>)
+      decreases height
+    {
+      forall e <- history :: f.requires(e.0) && e.1 == f(e.0)
+    }
+
+    method {:rlimit 0} Invoke(t: T) returns (r: R)
+      requires Requires(t)
+      reads Reads(t)
+      modifies Modifies(t)
+      decreases Decreases(t).Ordinal()
+      ensures Ensures(t, r)
+    {
+      assert Requires(t);
+      assert Valid();
+      r := f(t);
+      Update(t, r);
+
+      calc {
+        Produced();
+        old(Produced()) + [r];
+        old(Seq.MapPartialFunction(f, Consumed())) + [f(t)];
+        Seq.MapPartialFunction(f, old(Consumed())) + [f(t)];
+        Seq.MapPartialFunction(f, old(Consumed())) + Seq.MapPartialFunction(f, [t]);
+        { Seq.LemmaMapPartialFunctionDistributesOverConcat(f, old(Consumed()), [t]); }
+        Seq.MapPartialFunction(f, old(Consumed()) + [t]);
+        Seq.MapPartialFunction(f, Consumed());
+      }
+      assert Valid();
+    }
+
+    method RepeatUntil(t: T, stop: R -> bool, ghost eventuallyStopsProof: ProducesTerminatedProof<T, R>)
+      requires Valid()
+      requires eventuallyStopsProof.Action() == this
+      requires eventuallyStopsProof.FixedInput() == t
+      requires eventuallyStopsProof.StopFn() == stop
+      requires forall i <- Consumed() :: i == t
+      reads Repr
+      modifies Repr
+      decreases Repr
+      ensures Valid()
+    {
+      DefaultRepeatUntil(this, t, stop, eventuallyStopsProof);
+    }
+  }
+
+  // TODO: Move to Enumerators?
+  class FunctionalEnumerator<S, T> extends Action<(), Option<T>> {
+
+    const stepFn: S -> Option<(S, T)>
+    var state: S
+
+    ghost predicate Valid()
+      reads this, Repr
+      ensures Valid() ==> this in Repr
+      ensures Valid() ==> CanProduce(history)
+      decreases height, 0
+    {
+      this in Repr
+    }
+
+    constructor(state: S, stepFn: S -> Option<(S, T)>) {
+      this.state := state;
+      this.stepFn := stepFn;
+    }
+
+    ghost predicate CanConsume(history: seq<((), Option<T>)>, next: ())
+      decreases height
+    {
+      true
+    }
+    ghost predicate CanProduce(history: seq<((), Option<T>)>)
+      decreases height
+    {
+      true
+    }
+
+    method Invoke(t: ()) returns (r: Option<T>)
+      requires Requires(t)
+      reads Repr
+      modifies Modifies(t)
+      decreases Decreases(t).Ordinal()
+      ensures Ensures(t, r)
+    {
+      var next := stepFn(state);
+      match next {
+        case Some(result) => {
+          var (newState, result') := result;
```

Comment:
This looks nice. But if you prefer, it should also be possible to write it with a nested pattern:

``` dafny
case Some((newState, result')) =>
```

Also, note that the curly braces around an entire `case` body is never necessary, because each `case` is already a block scope by itself.